### DISCUSSION TITLE
.net 5 UserAccessor.User.GetIdentifier() cause error when login via openId

### DISCUSSION THREAD

**Discussion by minhhungit** [serenity-is/Serenity#5615] [Help] [answered]
- Serenity .net 5 version

When we login using openId (azure AD), [NameIdentifier](https://github.com/serenity-is/Serenity/blob/9e28e41ee4657d19ea57468a225390350285aef3/src/Serenity.Net.Core/Authorization/AuthorizationExtensions.cs#L68) claim returns "not-number" string so [Convert.ToInt32](https://github.com/serenity-is/Serene/blob/ea6e7581fa4043041d3db2b2bacab786a63d31ff/Serene/Serene.Core/Modules/Administration/User/Authentication/PermissionService.cs#L52) is not correct anymore

Normally value for `NameIdentifier `claim is from `Serenity.dbo.Users.UserID`

![image](https://user-images.githubusercontent.com/2279508/113757136-77487000-973c-11eb-8b86-fefb0fcc4422.png)

**Workarounds solution**: use `UserRetrieveService.ByUsername`

```csharp
var user = (UserDefinition)UserRetrieveService.ByUsername(username);
            
int userId = 0;
if (user != null)
{
    userId = int.Parse(user.Id);
}

if (userId == 0)
{
    userId = Convert.ToInt32(UserAccessor.User.GetIdentifier(), CultureInfo.InvariantCulture);
}
```

**Comment by VictorTomaili** [comment]
When you use different auth provider than serenity has own, you need implement own `PermissionService`, `UserRetrieveService` etc. if it's causes errors. You can replace them [there](https://github.com/serenity-is/Serene/blob/master/Serene/Serene.Core/Initialization/Startup.cs#L127)
Serenity has `Int32` id field and it's sets `Int32` value there while creating principal. So expected behavior is converting to Int32 not causes error. For custom implementations like yours, needs some extra implementations what currently doesn't included by default. 
OpenID implementation is welcome as pull request.

**Comment by minhhungit** [comment]
yes but problem is `GetIdentifier` method can be used in somewhere else and hardcode so I just worry that we can not use DI way.
Currently I just see we use the method in `PermissionService.cs` (sure we can implement a custom version of PermissionService and registry it using DI like you said), but not sure if in future we will use the method in somewhere. So I guess better way we should update `GetIdentifier` method

Do you think it's okay if we update the method like this:

```csharp
public static string GetIdentifier(this ClaimsPrincipal identity)
{
    if (identity == null)
        return null;

    // check IsLoggedIn
    if (!IsLoggedIn(identity))
        return null;
    
    // firstly, try to get user info by username
    var user = (TUserDefinition)userRetrieveService.ByUsername(identity.Identity.Name);
    if (user != null) 
    {
        return user.UserId;
    }

    // get userId from claim
    return identity.FindFirst(ClaimTypes.NameIdentifier)?.Value;
}
```

**Comment by VictorTomaili** [comment]
Are you point to [this](https://github.com/serenity-is/Serenity/blob/master/src/Serenity.Net.Core/Authorization/AuthorizationExtensions.cs#L68) method? 

This method doesn't interest about type of name identifier. It's always string there. It's implemented by `System.Security.Claims.Claim` class from Microsoft self. Not related to Serenity.

When you have ClaimsPrincipal, you must get identifier from there. Because It's already filled from database. Your code brings to one more query. Also both of value is returned as String and same value. 

Your exception occurs on `PermissionService` implementation, what expects `GetIdentifier` method returns an integer. 

I move this issue to Discussion.

**Reply by minhhungit** [reply]
Yes I mean to the [method](https://github.com/serenity-is/Serenity/blob/master/src/Serenity.Net.Core/Authorization/AuthorizationExtensions.cs#L68)

>This method doesn't interest about type of name identifier

And yes I agree that we will keep the method without updating it because it just does only one thing (just get value of NameIdentifier claim)

>Also both of value is returned as String and same value.

No, they are string but not same value.
With normal login, NameIdentifier return Serenity UserId while if we login via openId it returns a value which is not Serenity UserId (check red box in my image)

And because NameIdentifier claim can return a not-number string, [this line](https://github.com/serenity-is/Serene/blob/ea6e7581fa4043041d3db2b2bacab786a63d31ff/Serene/Serene.Core/Modules/Administration/User/Authentication/PermissionService.cs#L52) will leak an error. That value is also returned from microsoft, that mean we might be wrong to keep using Convert.ToInt32

It's okay to create a custom PermissionService class to replace current code for my project, just wonder should we update it for others users by updating Serene template code, then they don't need to write a custom class like me☺️

Because I am going to modify PermissionService  class for my project so I needed a small favor, this bellow code looks work but do you think I should use it with your knowledge about Serenity system:

I will update [this line](https://github.com/serenity-is/Serene/blob/ea6e7581fa4043041d3db2b2bacab786a63d31ff/Serene/Serene.Core/Modules/Administration/User/Authentication/PermissionService.cs#L52) with code:
```csharp
// try to get real Serenity userId by username before we get userId from UserAccessor.User.GetIdentifier()
var user = (UserDefinition)UserRetrieveService.ByUsername(username);
            
int userId = 0;
if (user != null)
{
    userId = int.Parse(user.Id);
}

// only when we can not get userId by username, we will try to get it from `NameIdentifier`
if (userId == 0)
{
    userId = Convert.ToInt32(UserAccessor.User.GetIdentifier(), CultureInfo.InvariantCulture);
}
```

**Reply by VictorTomaili** [reply]
Nope, your case is unique to you. This type of behavior must handled in PermissionService or related services. Not in a helper method what extract a value from an object. I said they are same value because It's returns same Id in serene application without OpenId implementation. When you implement your custom auth provider, you need create own services what cover your cases as well.

**Reply by minhhungit** [reply]
maybe you are confusing, I said above I will update PermissionService (update line Convert.ToInt32) instead GetIdentifier helper method, I did agree with you that we won't need to update GetIdentifier method :)

**Reply by VictorTomaili** [reply]
Hımm, If you can find a solution for check first user id instead UserRetrieveService it's may possible (TryParse etc.). But check first by username is not good way for this. Think about who really didn't need to get userId like you. There is a lot of project. I understand your implementation only gives error there, and you don't want to create a copy of this class. But I believe best way is still creating a separated class currently.

**Reply by minhhungit** [reply]
Okay @VictorTomaili , thank you for your help 👍

**Comment by minhhungit** [comment]
@VictorTomaili 

>Currently I just see we use the method in PermissionService.cs

Actually I was wrong, there are many class where we use `GetIdentifier` helper method 

![image](https://user-images.githubusercontent.com/2279508/113915463-07ec8200-9809-11eb-9220-61cedfe79ada.png)

**Comment by minhhungit** [comment]
@VictorTomaili @volkanceylan GetIdentifier() is used in many where, for example here https://github.com/serenity-is/Serenity/blob/9e28e41ee4657d19ea57468a225390350285aef3/src/Serenity.Net.Services/RequestHandlers/IntegratedFeatures/InsertUpdateLog/InsertUpdateLogBehavior.cs#L27

It is getting value not from table Serenity.dbo.User.Id, it's hardcode and I can not do anything

**Reply by minhhungit** [reply]
Problem is we treat GetIdentifier as "GetSerenityUserId", and this leak error when login via openId (maybe other login methods),  I suggest that we will create interface IGetSerenityUserIdHandler and inject it using DI 

for example 

```csharp
using System.Security.Claims;

namespace Serenity
{
    public interface IGetSerenityUserIdHandler
    {
        string GetUserId(ClaimsPrincipal identity);
    }
}

```

```csharp
public static class AuthorizationExtensions 
{
      private static IGetSerenityUserIdHandler SerenityUserIdHandler;

      public static void Configure(IGetSerenityUserIdHandler serenityUserIdHandler)
      {
          SerenityUserIdHandler = serenityUserIdHandler;
      }

      public static string GetIdentifier(this ClaimsPrincipal identity)
      {
          return SerenityUserIdHandler.GetUserId(identity);
      }

      // ....
}
```

In web project, Startup.cs/Configure

```csharp
AuthorizationExtensions.Configure(app.ApplicationServices.GetService<IGetSerenityUserIdHandler>());
```

@VictorTomaili @volkanceylan please review, if you guys okay with it I will create a pull request

**Comment by volkanceylan** [comment]
You shouldn't use the claims / ID etc. that you receive from OpenID / Azure / Google etc. as is, you should get the external user information and convert it to a cookie with claims that match the ones the application expect. I'm not willing to do any changes in the code at the moment, this works fine with many customers who use external logins.

**Comment by minhhungit** [comment, marked as answer]
Thank you @volkanceylan @VictorTomaili I found solution, I need to transform claim, for my case I used IClaimsTransformation, before this solution I have to modify and rebuild Serenity Dlls for myself :(

Post code here, it might help someone else

**Startup.cs**

```csharp
services.AddSingleton<IClaimsTransformation, Administration.MyClaimsTransformation>();
```

**And here is `MyClaimsTransformation`:**

```csharp
public class MyClaimsTransformation : IClaimsTransformation
{
    IUserRetrieveService UserRetrieveService;
    public MyClaimsTransformation(IUserRetrieveService userRetrieveService)
    {
        UserRetrieveService = userRetrieveService;
    }

    public Task<ClaimsPrincipal> TransformAsync(ClaimsPrincipal principal)
    {
        var identity = principal.Identities.FirstOrDefault(x => x.IsAuthenticated);
        if (identity == null) return Task.FromResult(principal);

        var nameIdenClaim = principal.Claims.FirstOrDefault(x => x.Type == ClaimTypes.NameIdentifier);
        if (nameIdenClaim != null)
        {
            principal.Identities.First().RemoveClaim(nameIdenClaim);
        }

        var username = principal.Claims.FirstOrDefault(x => x.Type == ClaimTypes.Name)?.Value;
        if (string.IsNullOrWhiteSpace(username))
        {
            throw new Exception("username is empty");
        }

        var userByUsername = UserRetrieveService.ByUsername(username);
        principal.Identities.First().AddClaim(new Claim(ClaimTypes.NameIdentifier, userByUsername.Id));

        return Task.FromResult(principal);
    }
}
```

**Reply by ysaadh** [reply]
Could you please share all steps that you made

**Reply by minhhungit** [reply]
@ysaadh I'm sorry this code relate to security problem so I guess I should not share it public. But, I can give example or links I think you can follow:

// Reference: https://joonasw.net/view/aspnet-core-2-azure-ad-authentication
// Complete sample application: https://github.com/juunas11/aspnetcore2aadauth
// Claims: https://andrewlock.net/introduction-to-authentication-with-asp-net-core/
